grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerProvider.java

```diff
@@ -0,0 +1,99 @@
+      CircuitBreakerRegistry circuitBreakerRegistry,
+      Map<String, CircuitBreakerConfig> circuitBreakerConfigMap,
+      List<String> disabledKeys) {
+    this.circuitBreakerRegistry = circuitBreakerRegistry;
+    this.circuitBreakerConfigMap = circuitBreakerConfigMap;
+    this.disabledKeys = disabledKeys;
+    this.defaultEnabled = !disabledKeys.contains(CircuitBreakerConfigParser.DEFAULT_THRESHOLDS);
+  }
+
+  public Optional<CircuitBreaker> getCircuitBreaker(String circuitBreakerKey) {
+    if (disabledKeys.contains(circuitBreakerKey)) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(
+        circuitBreakerCache.computeIfAbsent(
+            circuitBreakerKey,
+            key -> {
+              CircuitBreaker circuitBreaker =
+                  getCircuitBreakerFromConfigMap(circuitBreakerKey, defaultEnabled);
+              // If no circuit breaker is created return empty
+              if (circuitBreaker == null) {
+                return null; // Ensures cache does not store null entries
+              }
+              attachListeners(circuitBreaker);
+              return circuitBreaker;
+            }));
+  }
+
+  public Optional<CircuitBreaker> getSharedCircuitBreaker() {
+    if (!defaultEnabled) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        circuitBreakerCache.computeIfAbsent(
+            SHARED_KEY,
+            key -> {
+              CircuitBreaker circuitBreaker = circuitBreakerRegistry.circuitBreaker(SHARED_KEY);
+              attachListeners(circuitBreaker);
+              return circuitBreaker;
+            }));
+  }
+
+  private static void attachListeners(CircuitBreaker circuitBreaker) {
+    circuitBreaker
+        .getEventPublisher()
+        .onStateTransition(
+            event ->
+                log.info(
+                    "State transition: {} for circuit breaker {}",
+                    event.getStateTransition(),
+                    event.getCircuitBreakerName()))
+        .onCallNotPermitted(
+            event ->
+                log.debug(
+                    "Call not permitted: Circuit is OPEN for circuit breaker {}",
+                    event.getCircuitBreakerName()))
+        .onEvent(
+            event ->
+                log.debug(
+                    "Circuit breaker event type {} for circuit breaker name {}",
+                    event.getEventType(),
+                    event.getCircuitBreakerName()));
+  }
+
+  private CircuitBreaker getCircuitBreakerFromConfigMap(
+      String circuitBreakerKey, boolean defaultEnabled) {
+    return Optional.ofNullable(circuitBreakerConfigMap.get(circuitBreakerKey))
+        .map(config -> circuitBreakerRegistry.circuitBreaker(circuitBreakerKey, config))
+        .orElseGet(
+            () ->
+                defaultEnabled
+                    ? circuitBreakerRegistry.circuitBreaker(circuitBreakerKey)
+                    : null); // Return null if default is disabled
```

Comment:
Nit - given that most of this PR (including this same function) is working in optionals, please stick to optionals over null unless the third party code doesn't give us an option. This function can return an optional to convey it like
```java
private Optional<CircuitBreaker> buildNewCircuitBreaker(String circuitBreakerKey, boolean defaultEnabled) {
  return Optional.ofNullable(circuitBreakerConfigMap.get(circuitBreakerKey))
        .map(config -> circuitBreakerRegistry.circuitBreaker(circuitBreakerKey, config))
        .or(() -> defaultEnabled ? Optional.of(circuitBreakerRegistry.circuitBreaker(circuitBreakerKey)) : Optional.empty());
```

Reply:
fixed it